### ISSUE TITLE
Timezone support in DateTime()

### ISSUE THREAD

**Issue by abrander** [evanoberholster/imagemeta#26]
Timestamps are stored in an undefined timezone in EXIF. In `DateTime()` imagemeta assumes UTC. There is a note mentioning missing support for OffsetTimeOriginal and OffsetTimeDigitized, but I have yet to see those in the wild.

The current implementation assuming UTC is almost always wrong.

I have a few suggestions for fixing this.

1) Add a timezone argument to `DateTime()`. This will break API-compatibility and will push the problem to the user of the package.

2) Add an additional function like `DateTimeZone(tz *time.Location)`. This will provide a more correct function while maintaining backwards compatibility for using using `DateTime()`.

3) Add a `SetTimeZone()` method to `exif.Data`, and let it default to `time.UTC`. This will provide backwards compatibility and not expose another helper function.

4) Assume local timezone using `time.Local`. This will not provide backwards compatibility, and not really fix the problem, but it will probably guess correct more often than assuming UTC.

Would you be interested in a PR implementing one of those suggestions?

**Comment by evanoberholster** [comment]
Thanks for setting out those 4 options. 
- It's important that time zones are well represented, using only the local timezone or a set timezone would be insufficient.
- I have been working on a solution with [timezoneLookup](https://github.com/evanoberholster/timezoneLookup) to match latitude and longitude with timezone.
- I am ok with breaking changes in API at this stage if it moves toward a good standard, more intuitive/readable, and better performance.

Either option 1 or 2 would be feasible. Do you have other suggestions?

**Comment by abrander** [comment]
I'll like to add that I have some thoughts about GPS timestamps, I think it's outside the scope of this issue, but for completeness, I'll include it here. The EXIF GPS timestamp should represent the time in UTC, but I have observed the following:

- GPSTimestamp in UTC as expected.
- GPSTimestamp in UTC without accounting for leap seconds.
- GPSTimestamp in GPS time.
- GPSTimestamp in TAI (!).
- GPSTimestamp corrected for device time (!).
- GPSTimestamp representing the time of "location lock", not capture time.
- Strange errors around the first and last week of the year.

So in summary, not even the GPS timestamp can be used as any meaningful source of truth.

I think it's possible to infer a timezone based on coordinates (and a date). It would be useful if `timezoneLookup` provided a method returning a `time.Location` based on a coordinate (and a date) to easily interface with imagemeta.

I'll try to implement 1) as I think that makes the most sense, and it makes it very explicit that the package user has to make a choice.

**Comment by evanoberholster** [comment]
GPS timestamps are useful if you know what standard the cameras you are using have. Many people who travel often forget to update their date and time stamps on their DSLRs and Point and shoots.

Canon seems to use GPS timestamps based on GMT0

**Comment by abrander** [comment]
Well. Since #27 got merged, we're in a much better place, and I'll close this.

**Comment by mholt** [comment]
FWIW, Google Pixel phones use OffsetTime, OffsetTimeOriginal, and OffsetTimeDigitized for time zone information. Looks like the value is, e.g. `-04:00` but I don't know if that's the "raw" value in the file or if that's an interpreted value. (Multiple exif tools have showed this value.)

The current exif lib I'm using, cozy/goexif2, doesn't read those. Will/does this package support those fields?

**Comment by evanoberholster** [comment]
@mholt thanks for mentioning this. It appears to be the raw value. 7 bytes long ending in a null byte.

I have just pushed an update to ```develop``` branch that includes these 3 Tags. Due to the way ```time.Location``` works I am struggling to convert ```-04:00``` to a Location to be included in a ```time.Time```. Have you been able to do this?

**Comment by mholt** [comment]
@evanoberholster Thanks for the reply!

Yep, I can confirm. I also found this confusing but after a bit of fiddling, you can see what I got in the linked PR above:

```go
// timeOffsetVal is the EXIF field value as a string string, e.g. "-04:00"
t, err := time.Parse("-07:00", timeOffsetVal)
if err != nil {
	return nil, err
}
return t.Location(), nil
```

Hope that helps :) Thanks so much again for working on this lib. Really excited for it.

**Comment by evanoberholster** [comment]
Hi @mholt, thanks for contributing ideas. I made some of the changes you asked for. Took a deeper look at the ```time``` package to find the most efficient way to implement offsets. I still think it could be improved, right now it has 4 allocations per tag.

Take a look at 5fb96b9. Kindly fork the ```develop``` branch when you get the chance and test it.

I hope to soon be able to push ```develop``` to ```main``` after the rewrite. That way PRs can be smaller with changes instead of major rewrites.

**Comment by mholt** [comment]
I went ahead and tested the timestamp values from the `develop` branch to make sure the timezone is properly applied, and sure enough it looks correct. Thank you so much!

**Comment by evanoberholster** [comment]
Closing this issue as it appears resolved. Please reopen if it needs to be further addressed.
